File: java/src/main/java/com/penapereira/patterns/producerconsumer/BoundedBuffer.java
package com.penapereira.patterns.producerconsumer;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * A monitor: one lock, a "not full" and a "not empty" condition. This is what
 * java.util.concurrent.ArrayBlockingQueue does internally.
 */
public final class BoundedBuffer<T> {

    private final int capacity;
    private final Deque<T> items = new ArrayDeque<>();
    private final ReentrantLock lock = new ReentrantLock();
    private final Condition notFull = lock.newCondition();
    private final Condition notEmpty = lock.newCondition();

    public BoundedBuffer(int capacity) {
        this.capacity = capacity;
    }

    /** Waits while the buffer is full, then appends the item. */
    public void put(T item) throws InterruptedException {
        lock.lock();
        try {
            while (items.size() == capacity) {
                notFull.await();
            }
            items.addLast(item);
            notEmpty.signal();
        } finally {
            lock.unlock();
        }
    }

    /** Waits while the buffer is empty, then removes the oldest item. */
    public T take() throws InterruptedException {
        lock.lock();
        try {
            while (items.isEmpty()) {
                notEmpty.await();
            }
            T item = items.removeFirst();
            notFull.signal();
            return item;
        } finally {
            lock.unlock();
        }
    }

    public int size() {
        lock.lock();
        try {
            return items.size();
        } finally {
            lock.unlock();
        }
    }
}

File: python/src/patterns/producerconsumer/bounded_buffer.py
import threading
from collections import deque


class BoundedBuffer[T]:
    """A monitor: one lock (inside the Condition) and two wait conditions.

    ``queue.Queue(maxsize)`` is the standard-library equivalent.
    """

    def __init__(self, capacity: int) -> None:
        self._capacity = capacity
        self._items: deque[T] = deque()
        self._lock = threading.Lock()
        self._not_full = threading.Condition(self._lock)
        self._not_empty = threading.Condition(self._lock)

    def put(self, item: T) -> None:
        """Wait while the buffer is full, then append the item."""
        with self._not_full:
            self._not_full.wait_for(lambda: len(self._items) < self._capacity)
            self._items.append(item)
            self._not_empty.notify()

    def take(self) -> T:
        """Wait while the buffer is empty, then remove the oldest item."""
        with self._not_empty:
            self._not_empty.wait_for(lambda: len(self._items) > 0)
            item = self._items.popleft()
            self._not_full.notify()
            return item

    def size(self) -> int:
        with self._lock:
            return len(self._items)
